File: src/lib/auth-service.js
import { NextRequest, NextResponse } from 'next/server';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { getMongoClient } from './mongodb';
import { USERS_COLLECTION } from './constants';

// JWT Configuration
const JWT_SECRET_KEY = process.env.JWT_SECRET_KEY || "your-secret-key-change-in-production-2024";
const JWT_ALGORITHM = "HS256";
const TOKEN_EXPIRY_HOURS = 24;

// Collections
const USERS_AUTH_COLLECTION = "authenticated_users";
const USER_SESSIONS_COLLECTION = "user_sessions";

class AuthService {
  constructor() {
    this.db = null;
    this.usersCollection = null;
    this.sessionsCollection = null;
    this.kycCollection = null;
  }

  async initialize() {
    if (!this.db) {
      this.db = await getMongoClient();
      this.usersCollection = this.db.collection(USERS_AUTH_COLLECTION);
      this.sessionsCollection = this.db.collection(USER_SESSIONS_COLLECTION);
      this.kycCollection = this.db.collection(USERS_COLLECTION);
    }
  }

  hashPassword(password) {
    const salt = bcrypt.genSaltSync(10);
    return bcrypt.hashSync(password, salt);
  }

  verifyPassword(password, hashedPassword) {
    try {
      return bcrypt.compareSync(password, hashedPassword);
    } catch (error) {
      console.error(`DEBUG: Error verifying password: ${error}`);
      return false;
    }
  }

  async registerUser(kycData) {
    try {
      await this.initialize();

      // Check if user already exists
      const existingUser = await this.usersCollection.findOne({
        $or: [
          { email: kycData.email },
          { username: kycData.email }
        ]
      });

      if (existingUser) {
        console.log("DEBUG: User already exists");
        return null;
      }

      // Hash the password
      const hashedPassword = this.hashPassword(kycData.password);

      // Create user document
      const userDoc = {
        username: kycData.email,
        email: kycData.email,
        password_hash: hashedPassword,
        name: kycData.name,
        mobile: kycData.mobile,
        faculty: kycData.faculty,
        role: "user",
        is_active: true,
        created_at: new Date(),
        last_login: null,
        session_id: kycData.sessionId || "unknown"
      };

      const result = await this.usersCollection.insertOne(userDoc);
      const userId = result.insertedId.toString();

      console.log(`DEBUG: Successfully registered user with ID: ${userId}`);
      return userId;

    } catch (error) {
      console.error(`DEBUG: Error registering user: ${error}`);
      return null;
    }
  }

  async authenticateUser(username, password) {
    try {
      await this.initialize();

      // Find user by username or email
      const user = await this.usersCollection.findOne({
        $or: [
          { username: username.toLowerCase() },
          { email: username.toLowerCase() }
        ],
        is_active: true
      });

      if (!user) {
        console.log(`DEBUG: User not found: ${username}`);
        return [null, null];
      }

      // Verify password
      if (!this.verifyPassword(password, user.password_hash)) {
        console.log(`DEBUG: Invalid password for user: ${username}`);
        return [null, null];
      }

      // Update last login
      await this.usersCollection.updateOne(
        { _id: user._id },
        { $set: { last_login: new Date() } }
      );

      // Generate JWT token
      const token = this.generateJwtToken(user);

      // Remove sensitive data from user object
      const userData = {
        user_id: user._id.toString(),
        username: user.username,
        email: user.email,
        name: user.name,
        mobile: user.mobile,
        faculty: user.faculty,
        role: user.role || "user",
        last_login: user.last_login
      };

      console.log(`DEBUG: Successfully authenticated user: ${username}`);
      return [token, userData];

    } catch (error) {
      console.error(`DEBUG: Error authenticating user: ${error}`);
      return [null, null];
    }
  }

  generateJwtToken(user) {
    try {
      // Create payload
      const payload = {
        user_id: user._id.toString(),
        username: user.username,
        email: user.email,
        role: user.role || "user",
        iat: Math.floor(Date.now() / 1000),
        exp: Math.floor(Date.now() / 1000) + (TOKEN_EXPIRY_HOURS * 3600)
      };

      return jwt.sign(payload, JWT_SECRET_KEY, { 
        algorithm: JWT_ALGORITHM,
        expiresIn: `${TOKEN_EXPIRY_HOURS}h`
      });

    } catch (error) {
      console.error(`DEBUG: Error generating JWT token: ${error}`);
      return null;
    }
  }

  async verifyToken(token) {
    try {
      const decoded = jwt.verify(token, JWT_SECRET_KEY, { 
        algorithms: [JWT_ALGORITHM] 
      });

      await this.initialize();

      // Check if user still exists and is active
      const user = await this.usersCollection.findOne({
        _id: decoded.user_id,
        is_active: true
      });

      if (!user) {
        console.log(`DEBUG: User not found or inactive for token: ${decoded.user_id}`);
        return null;
      }

      // Check if session is still valid (optional - check sessions collection)
      const session = await this.sessionsCollection.findOne({
        user_id: decoded.user_id,
        token: token,
        expires_at: { $gt: new Date() }
      });

      return {
        user_id: decoded.user_id,
        username: decoded.username,
        email: decoded.email,
        role: decoded.role,
        session_valid: !!session
      };

    } catch (error) {
      if (error.name === 'TokenExpiredError') {
        console.log("DEBUG: Token expired");
      } else if (error.name === 'JsonWebTokenError') {
        console.log(`DEBUG: Invalid token: ${error.message}`);
      } else {
        console.error(`DEBUG: Error verifying token: ${error}`);
      }
      return null;
    }
  }

  async logoutUser(token) {
    try {
      await this.initialize();

      const decoded = jwt.verify(token, JWT_SECRET_KEY);
      
      // Deactivate session
      await this.sessionsCollection.updateOne(
        { user_id: decoded.user_id, token: token },
        { $set: { is_active: false, logged_out_at: new Date() } }
      );

      return true;

    } catch (error) {
      console.error(`DEBUG: Error logging out user: ${error}`);
      return false;
    }
  }

  async getUserById(userId) {
    try {
      await this.initialize();

      const user = await this.usersCollection.findOne({
        _id: userId,
        is_active: true
      });

      if (user) {
        // Remove sensitive data
        const { password_hash, ...userData } = user;
        return userData;
      }
      return null;

    } catch (error) {
      console.error(`DEBUG: Error getting user by ID: ${error}`);
      return null;
    }
  }

  async cleanupExpiredSessions() {
    try {
      await this.initialize();

      await this.sessionsCollection.deleteMany({
        expires_at: { $lt: new Date() }
      });

      console.log("DEBUG: Expired sessions cleaned up");

    } catch (error) {
      console.error(`DEBUG: Error cleaning up expired sessions: ${error}`);
    }
  }
}

// Global auth service instance
const authService = new AuthService();

export { authService };

File: oldapp/auth_service.py
import jwt
import bcrypt
import hashlib
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
import chainlit as cl
from mongo_util import get_mongo_client
from constants import USERS_COLLECTION

# JWT Configuration - In production, these should be environment variables
JWT_SECRET_KEY = "your-secret-key-change-in-production-2024"  # Change this in production
JWT_ALGORITHM = "HS256"
TOKEN_EXPIRY_HOURS = 24

# Collections
USERS_AUTH_COLLECTION = "authenticated_users"
USER_SESSIONS_COLLECTION = "user_sessions"

class AuthService:
    def __init__(self):
        self.db = get_mongo_client()
        self.users_collection = self.db[USERS_AUTH_COLLECTION]
        self.sessions_collection = self.db[USER_SESSIONS_COLLECTION]
        self.kyc_collection = self.db[USERS_COLLECTION]  # KYC collection for data migration
    
    def hash_password(self, password: str) -> str:
        """Hash password using bcrypt"""
        # Convert password to bytes and hash
        password_bytes = password.encode('utf-8')
        salt = bcrypt.gensalt()
        hashed = bcrypt.hashpw(password_bytes, salt)
        return hashed.decode('utf-8')
    
    def verify_password(self, password: str, hashed_password: str) -> bool:
        """Verify password against hash"""
        try:
            password_bytes = password.encode('utf-8')
            hashed_bytes = hashed_password.encode('utf-8')
            return bcrypt.checkpw(password_bytes, hashed_bytes)
        except Exception as e:
            print(f"DEBUG: Error verifying password: {e}")
            return False
    
    def register_user(self, kyc_data: Dict[str, Any]) -> Optional[str]:
        """Register new user from KYC data"""
        try:
            # Check if user already exists
            existing_user = self.users_collection.find_one({
                "$or": [
                    {"email": kyc_data.get("email")},
                    {"username": kyc_data.get("email")}  # Use email as username initially
                ]
            })
            
            if existing_user:
                print(f"DEBUG: User already exists with email: {kyc_data.get('email')}")
                return None
            
            # Hash the password
            hashed_password = self.hash_password(kyc_data.get("password"))
            
            # Create user document
            user_doc = {
                "username": kyc_data.get("email"),  # Use email as username
                "email": kyc_data.get("email"),
                "password_hash": hashed_password,
                "name": kyc_data.get("name"),
                "mobile": kyc_data.get("mobile"),
                "faculty": kyc_data.get("faculty"),
                "role": "user",  # Default role
                "is_active": True,
                "created_at": datetime.utcnow(),
                "last_login": None,
                "session_id": cl.user_session.get("id", "unknown"),  # Link to original KYC session
            }
            
            result = self.users_collection.insert_one(user_doc)
            user_id = str(result.inserted_id)
            
            print(f"DEBUG: Successfully registered user with ID: {user_id}")
            return user_id
            
        except Exception as e:
            print(f"DEBUG: Error registering user: {e}")
            return None
    
    def authenticate_user(self, username: str, password: str) -> tuple[Optional[str], Optional[Dict]]:
        """Authenticate user and return JWT token and user data"""
        try:
            # Find user by username or email
            user = self.users_collection.find_one({
                "$or": [
                    {"username": username.lower()},
                    {"email": username.lower()}
                ],
                "is_active": True
            })
            
            if not user:
                print(f"DEBUG: User not found: {username}")
                return None, None
            
            # Verify password
            if not self.verify_password(password, user["password_hash"]):
                print(f"DEBUG: Invalid password for user: {username}")
                return None, None
            
            # Update last login
            self.users_collection.update_one(
                {"_id": user["_id"]},
                {"$set": {"last_login": datetime.utcnow()}}
            )
            
            # Generate JWT token
            token = self.generate_jwt_token(user)
            
            # Remove sensitive data from user object
            user_data = {
                "user_id": str(user["_id"]),
                "username": user["username"],
                "email": user["email"],
                "name": user.get("name"),
                "mobile": user.get("mobile"),
                "faculty": user.get("faculty"),
                "role": user.get("role", "user"),
                "last_login": user.get("last_login")
            }
            
            print(f"DEBUG: Successfully authenticated user: {username}")
            return token, user_data
            
        except Exception as e:
            print(f"DEBUG: Error authenticating user: {e}")
            return None, None
    
    def generate_jwt_token(self, user: Dict[str, Any]) -> str:
        """Generate JWT token for authenticated user"""
        try:
            # Create payload
            payload = {
                "user_id": str(user["_id"]),
                "username": user["username"],
                "email": user["email"],
                "role": user.get("role", "user"),
                "exp": datetime.utcnow() + timedelta(hours=TOKEN_EXPIRY_HOURS),
                "iat": datetime.utcnow(),
                "session_id": cl.user_session.get("id", "unknown")
            }
            
            # Generate token
            token = jwt.encode(payload, JWT_SECRET_KEY, algorithm=JWT_ALGORITHM)
            
            # Store session in MongoDB
            session_doc = {
                "user_id": str(user["_id"]),
                "token": token,
                "session_id": cl.user_session.get("id", "unknown"),
                "created_at": datetime.utcnow(),
                "expires_at": payload["exp"],
                "is_active": True,
                "user_agent": "Chainlit-ChatBot",  # Could be enhanced with actual user agent
            }
            
            # Clean up old sessions for this user (keep only last 5)
            old_sessions = list(self.sessions_collection.find(
                {"user_id": str(user["_id"])},
                sort=[("created_at", -1)],
                skip=4  # Keep 4 most recent, delete the rest
            ))
            
            if old_sessions:
                old_session_ids = [session["_id"] for session in old_sessions]
                self.sessions_collection.delete_many({"_id": {"$in": old_session_ids}})
            
            # Insert new session
            self.sessions_collection.insert_one(session_doc)
            
            print(f"DEBUG: Generated JWT token for user: {user['username']}")
            return token
            
        except Exception as e:
            print(f"DEBUG: Error generating JWT token: {e}")
            return None
    
    def verify_token(self, token: str) -> Optional[Dict[str, Any]]:
        """Verify JWT token and return user data"""
        try:
            # Decode token
            payload = jwt.decode(token, JWT_SECRET_KEY, algorithms=[JWT_ALGORITHM])
            
            # Check if session exists and is active
            session = self.sessions_collection.find_one({
                "token": token,
                "is_active": True,
                "expires_at": {"$gt": datetime.utcnow()}
            })
            
            if not session:
                print(f"DEBUG: Session not found or expired for token")
                return None
            
            # Get current user data
            user = self.users_collection.find_one({
                "_id": session["user_id"],
                "is_active": True
            })
            
            if not user:
                print(f"DEBUG: User not found or inactive: {session['user_id']}")
                return None
            
            # Return user data
            user_data = {
                "user_id": str(user["_id"]),
                "username": user["username"],
                "email": user["email"],
                "name": user.get("name"),
                "mobile": user.get("mobile"),
                "faculty": user.get("faculty"),
                "role": user.get("role", "user"),
                "session_id": payload.get("session_id"),
                "token": token
            }
            
            return user_data
            
        except jwt.ExpiredSignatureError:
            print(f"DEBUG: Token expired")
            return None
        except jwt.InvalidTokenError as e:
            print(f"DEBUG: Invalid token: {e}")
            return None
        except Exception as e:
            print(f"DEBUG: Error verifying token: {e}")
            return None
    
    def logout_user(self, token: str) -> bool:
        """Logout user by deactivating token"""
        try:
            # Deactivate session
            result = self.sessions_collection.update_one(
                {"token": token},
                {"$set": {"is_active": False, "logout_at": datetime.utcnow()}}
            )
            
            success = result.modified_count > 0
            print(f"DEBUG: Logout {'successful' if success else 'failed'}")
            return success
            
        except Exception as e:
            print(f"DEBUG: Error logging out user: {e}")
            return False
    
    def get_user_by_id(self, user_id: str) -> Optional[Dict[str, Any]]:
        """Get user data by user ID"""
        try:
            user = self.users_collection.find_one({"_id": user_id, "is_active": True})
            if user:
                return {
                    "user_id": str(user["_id"]),
                    "username": user["username"],
                    "email": user["email"],
                    "name": user.get("name"),
                    "mobile": user.get("mobile"),
                    "faculty": user.get("faculty"),
                    "role": user.get("role", "user"),
                }
            return None
        except Exception as e:
            print(f"DEBUG: Error getting user by ID: {e}")
            return None
    
    def cleanup_expired_sessions(self):
        """Clean up expired sessions"""
        try:
            result = self.sessions_collection.delete_many({
                "expires_at": {"$lt": datetime.utcnow()}
            })
            print(f"DEBUG: Cleaned up {result.deleted_count} expired sessions")
            return result.deleted_count
        except Exception as e:
            print(f"DEBUG: Error cleaning up sessions: {e}")
            return 0

# Global auth service instance
auth_service = AuthService()
